self-contained/bootstrap_config.sh

```diff
@@ -386,6 +386,30 @@ ensureDHCPv6() {
 ensureKubelet() {
     KUBELET_DEFAULT_FILE=/etc/default/kubelet
     mkdir -p /etc/default
+
+    # In k8s >= 1.29 kubelet no longer sets node internalIP when using external cloud provider
+    # https://github.com/kubernetes/kubernetes/pull/121028
+    # This regresses node startup performance in Azure CNI Overlay and Podsubnet clusters, which require the node to be
+    # assigned an internal IP before configuring pod networking.
+    # To improve node startup performance, explicitly set `--node-ip` to the IP returned from IMDS so kubelet sets
+    # the internal IP when it registers the node.
+    # If this fails, skip setting --node-ip, which is safe because cloud-node-manager will assign it later anyway.
+    if semverCompare ${KUBERNETES_VERSION:-"0.0.0"} "1.29.0"; then
+        imdsOutput=$(curl -s -H Metadata:true --noproxy "*" --max-time 5 "http://169.254.169.254/metadata/instance?api-version=2021-02-01" 2> /dev/null)
+        if [[ $? -eq 0 ]]; then
+            nodeIPAddrs=()
+            ipv4Addr=$(echo $imdsOutput | jq -r '.network.interface[0].ipv4.ipAddress[0].privateIpAddress // ""')
+            [ -n "$ipv4Addr" ] && nodeIPAddrs+=("$ipv4Addr")
+            ipv6Addr=$(echo $imdsOutput | jq -r '.network.interface[0].ipv6.ipAddress[0].privateIpAddress // ""')
+            [ -n "$ipv6Addr" ] && nodeIPAddrs+=("$ipv6Addr")
+            nodeIPArg=$(IFS=, ; echo "${nodeIPAddrs[*]}") # join, comma-separated
+            if [ -n "$nodeIPArg" ]; then
+                echo "Adding --node-ip=$nodeIPArg to kubelet flags"
+                KUBELET_FLAGS="$KUBELET_FLAGS --node-ip=$nodeIPArg"
+            fi
+        fi
+    fi
```

Comment:
I copy-pasted this here. Is that the right way to do this?

Reply:
cc @Devinwong @lilypan26 

I think that's fine for now... I wonder if we should add some automation which would get invoked by `make generate` to copy relevant changes over to `self-contained`

I know we have similar logic for various things in the RP repo

Reply:
that makes sense. Do E2E tests cover k8s 1.29+ with self-contained VHD? want to make sure this doesn't regress that configuration.